src/duration/mod.rs

```diff
@@ -0,0 +1,356 @@
+        }
+
+        let mut millis = self.0.as_millis();
+        let mut formatted = String::new();
+
+        for unit in DurationUnit::iter() {
+            let whole = millis / unit.millis();
+            let rest = millis % unit.millis();
+
+            if whole > 0 {
+                write!(formatted, "{}{} ", whole, unit)?;
+            }
+
+            millis = rest;
+        }
+
+        write!(f, "{}", formatted.trim_end())
+    }
+}
+
+impl Deref for Duration {
+    type Target = std::time::Duration;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for Duration {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl From<std::time::Duration> for Duration {
+    fn from(value: std::time::Duration) -> Self {
+        Self(value)
+    }
+}
+
+impl Duration {
+    /// Creates a new [`Duration`] from the specified number of whole seconds.
+    pub const fn from_secs(secs: u64) -> Self {
+        Self(std::time::Duration::from_secs(secs))
+    }
+
+    /// Creates a new [`Duration`] from the specified number of whole
+    /// milliseconds.
+    pub const fn from_millis(millis: u64) -> Self {
+        Self(std::time::Duration::from_millis(millis))
+    }
+}
+
+/// Defines supported [`DurationUnit`]s. Each [`DurationFragment`] consists of
+/// a numeric value followed by a [`DurationUnit`]. The order of variants
+/// **MATTERS**. It is the basis for the correct transformation of the
+/// [`std::time::Duration`] back to a human-readable format, which is defined
+/// in the [`Display`] implementation of [`Duration`].
+#[derive(Debug, strum::EnumString, strum::Display, strum::AsRefStr, strum::EnumIter)]
+pub enum DurationUnit {
+    #[strum(serialize = "d")]
+    Days,
+
+    #[strum(serialize = "h")]
+    Hours,
+
+    #[strum(serialize = "m")]
+    Minutes,
+
+    #[strum(serialize = "s")]
+    Seconds,
+
+    #[strum(serialize = "ms")]
+    Milliseconds,
+}
+
+impl DurationUnit {
+    /// Returns the number of whole milliseconds in each supported
+    /// [`DurationUnit`].
+    pub fn millis(&self) -> u128 {
+        use DurationUnit::*;
+
+        match self {
+            Days => 24 * Hours.millis(),
+            Hours => 60 * Minutes.millis(),
+            Minutes => 60 * Seconds.millis(),
+            Seconds => 1000,
+            Milliseconds => 1,
+        }
+    }
+}
+
+#[derive(Debug, Error, PartialEq)]
+pub enum DurationFragmentParseError {
+    #[error("invalid input, either empty or contains non-ascii characters")]
+    InvalidInput,
+
+    #[error("expected number, the duration fragment must start with a numeric character")]
+    ExpectedNumber,
+
+    #[error("expected character, the duration fragments must end with an alphabetic character")]
+    ExpectedCharacter,
+
+    #[error("failed to parse fragment value as integer")]
+    ParseIntError(#[from] ParseIntError),
+
+    #[error("failed to parse fragment unit")]
+    UnitParseError,
```

Comment:
These error messages could use some context on where exactly the problem is.

Reply:
Agreed, I will add some more context.

Reply:
Added in 9e9257d